### DISCUSSION TITLE
how to access env variable in sass?

### DISCUSSION THREAD

**Discussion by naveennsit** [vercel/next.js#19042] [Help] [answered]
could you please tell me how to access `env` variable in `scss`.Here is my code

https://repl.it/@naveennsit/DirtyAcrobaticClasslibrary

I am able to access `env` variable in `jsx` like that

` <a>About sss{process.env.customKey}</a>`

I created `env` variable like this

    module.exports = {
      env: {
        customKey: "my-value",
        red: "red",
        bg:'data:image/jpeg;base64,

Now I want to access `variable` in scss file.But I am getting error.

I am trying to access like this
 

    // background-image: url($process.env.bg})

**Comment by mattcarlotta** [comment, marked as answer]
This isn't going to be easy and straightforward since [additionalData](https://github.com/webpack-contrib/sass-loader#additionaldata) isn't exposed by Next...

In short, you can think of `additionalData` as appending a string to the top of ALL of your sass files.

```scss
$image: "https://example.com/12345.jpg";
```

To do that, you're going to have manually add an `additionalData` property to Next's webpack `sass-loader` rule:

**next.config.js**
```js
const { NODE_ENV } = process.env;

const image = `https://via.placeholder.com/600/${NODE_ENV === "development" ? "92c952" : "771796"}`

const regexEqual = (x, y) =>
  x instanceof RegExp &&
  y instanceof RegExp &&
  x.source === y.source &&
  x.global === y.global &&
  x.ignoreCase === y.ignoreCase &&
  x.multiline === y.multiline;

module.exports = {
  webpack(config) {
    /* find all scss/sass rules */
    const sassRules = config.module.rules
      .find((rule) => typeof rule.oneOf === "object")
      .oneOf.find(
        (rule) =>
          rule.sideEffects === false &&
          regexEqual(rule.test, /\.module\.(scss|sass)$/)
      );

    /* map over style rules and add "additionalData" only to the "sass-loader" rule*/
    sassRules.use = sassRules.use.map((rule) =>
      rule.loader.includes("sass-loader")
        ? {
            ...rule,
            options: {
              ...rule.options,
              additionalData: `$image: '${image}';`,
            },
          }
        : rule
    );

    /* return new config to next */
    return config;
  },
};
```

Now in a `scss` or `sass` file, you can reference the variable `$image` (using modules or globals -- this example uses modules):
**components/HelloWorld/HelloWorld.module.scss**
```scss
$color: red;

.hello {
  color: $color;
}

.image {
  height: 600px;
  width: 600px;
  background-image: url($image);
}
```

And then apply/reference the class name:
**components/Hello/index.js**
```jsx
import styles from "./HelloWorld.module.scss";

export default function HelloWorld() {
  return (
    <div className={styles.hello}>
      Hello World, I am being styled using SCSS Modules!
      <div className={styles.image} />
    </div>
  );
}
```
---

On a related note, you may need multiple environment specific variables which can be quite cumbersome to add to `additionalData`. Instead, I'd recommend [importing](https://sass-lang.com/documentation/at-rules/import) a specific [partial](https://sass-lang.com/guide#topic-4) file like: `@import "vars.development.scss";` or `@import "vars.production.scss";`. 

Here's a work demo of importing partials based upon `NODE_ENV`: [Demo](https://github.com/mattcarlotta/with-next-sass-additional-data)

**Reply by naveennsit** [reply]
Thanks can you look into this question also https://github.com/vercel/next.js/discussions/19043

**Comment by tsfahmad** [comment]
Late to answer but now in NextJS 12 you can achieve the same by simply writing `sassOptions` in next.config
```
sassOptions: {
    prependData: `$image: '${image}';`,
  },
```
Define you image variable in the next.config file
`const image = `https://via.placeholder.com/600/${NODE_ENV === "development" ? "92c952" : "771796"}`
`

**Reply by GabenGar** [reply]
What is the shape of `sassOptions` anyway? `prependData` is not a property of [`sass-loader`](https://github.com/webpack-contrib/sass-loader#additionaldata) and neither of [SASS API options](https://sass-lang.com/documentation/js-api/interfaces/Options).

**Reply by GabenGar** [reply]
Also it doesn't work with SASS modules, since it crashes with `undefined variable` error.

**Reply by q7y331xk** [reply]
you may try prependData->additionalData 
it depends on your sass version
it work for me

**Reply by q7y331xk** [reply]
prependData is old version variable name and changed to addtionalData

**Reply by yanis-ouledmoussa** [reply]
Thanks alot ! Had hard time to finally find the right answer !!
